RandBLAS/base.hh

```diff
@@ -1,131 +1,191 @@
+        ro[2*i    ] = v0;
+        ro[2*i + 1] = v1;
     }
+    return ro;
 }
 
-// convert from RNGState to Random123_RNGState
-template <typename T_gen>
-Random123_RNGState<T_gen>::Random123_RNGState(
-    const RNGState &s
-) : ctr{},
-    key{},
-    len_c(T_gen::ctr_type::static_size),
-    len_k(T_gen::key_type::static_size)
+/// Generate a sequence of random values and apply a Box-Muller transform.
+struct boxmul
 {
-    bool res = generator_type_is_same<T_gen>(s);
-
-    if (!res) {
-        throw std::runtime_error(std::string("T_gen must match s.rng_name."));
+    /** Generate a sequence of random values and apply a Box-Muller transform.
+     *
+     * @tparam RNG a random123 CBRNG type
+     * @tparam T the desired return type. The default return type is dictated by
+     *           the RNG's counter element type. float for 32 bit counter elements
+     *           and double for 64.
+     *
+     * @param[in] a random123 CBRNG instance used to generate the sequence
+     * @param[in] the CBRNG counter of n elements
+     * @param[in] the CBRNG key
+     *
+     * @returns the generated and transformed sequence of n floating point
+     *          values. The return type is dictated by the RNG's counter
+     *          element type. float for 32 bit counter elements and double for
+     *          64.
+     */
+    template <typename RNG>
+    static
+    auto generate(
+        RNG &rng,
+        typename RNG::ctr_type const& c,
+        typename RNG::key_type const& k
+    ) {
+        return boxmulall<RNG>(rng(c,k));
     }
+};
 
-    int ctr_len = std::min(this->len_c, s.len_c);
-    memcpy(this->ctr.v, s.ctr, sizeof(uint32_t) * ctr_len);
-
-    int key_len = std::max(this->len_k, s.len_k);
-    memcpy(this->key.v, s.key, sizeof(uint32_t) * key_len);
-}
-
+/// Generate a sequence of random values and transform to -1.0 to 1.0.
+struct uneg11
+{
+    /** Generate a sequence of random values and transform to -1.0 to 1.0.
+     *
+     * @tparam RNG a random123 CBRNG type
+     * @tparam T the desired return type. The default return type is dictated by
+     *           the RNG's counter element type. float for 32 bit counter elements
+     *           and double for 64.
+     *
+     * @param[in] a random123 CBRNG instance used to generate the sequence
+     * @param[in] the CBRNG counter of n elements
+     * @param[in] the CBRNG key
+     *
+     * @returns the generated and transformed sequence of n floating point
+     *          values. The return type is dictated by the counter element
+     *          type. float for 32 bit counter elements and double for 64.
+     */
+    template <typename RNG, typename T = typename std::conditional
+        <sizeof(typename RNG::ctr_type::value_type) == sizeof(uint32_t), float, double>::type>
+    static
+    auto generate(
+        RNG &rng,
+        typename RNG::ctr_type const& c,
+        typename RNG::key_type const& k
+    ) {
+        return r123::uneg11all<T>(rng(c,k));
```

Comment:
I can't actually infer the return type here. My preference is for the return type to be explicit rather than ``auto``. If that's a big pain then I'd like to have some examples of return types in the docstring.

Reply:
it will be a messy template in this case. It is documented in the comment, but it may not be clear where n comes from. It's, the counter length, or 4 in Philox4x32. Let me see if I can make the comment easier to understand.

Reply:
resolved in c9303c27bb08